omp/preconditioner/isai_kernels.cpp

```diff
@@ -240,14 +248,115 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_ISAI_GENERATE_TRI_INVERSE_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+inline IndexType choose_pivot(IndexType block_size, const ValueType *block,
+                              size_type stride)
+{
+    IndexType cp = 0;
+    for (IndexType i = 1; i < block_size; ++i) {
+        if (abs(block[cp * stride]) < abs(block[i * stride])) {
+            cp = i;
+        }
+    }
+    return cp;
+}
+
+
+template <typename ValueType, typename IndexType>
+inline void swap_rows(IndexType row1, IndexType row2, IndexType block_size,
+                      ValueType *block, size_type stride)
+{
+    using std::swap;
+    for (IndexType i = 0; i < block_size; ++i) {
+        swap(block[row1 * stride + i], block[row2 * stride + i]);
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void generate_general_inverse(std::shared_ptr<const DefaultExecutor> exec,
+                              const matrix::Csr<ValueType, IndexType> *mtx,
+                              matrix::Csr<ValueType, IndexType> *inverse_mtx,
+                              IndexType *excess_rhs_ptrs,
+                              IndexType *excess_nz_ptrs, bool spd)
+{
+    using std::swap;
+    auto general_solve =
+        [spd](const range<accessor::row_major<ValueType, 2>> system,
+              ValueType *rhs, const IndexType rhs_one_idx) {
+            const IndexType size = system.length(0);
+            if (size <= 0) {
+                return;
+            }
+            // RHS is the identity: zero everywhere except for the diagonal
+            // entry
+            std::fill_n(rhs, size, zero<ValueType>());
+            rhs[rhs_one_idx] = one<ValueType>();
+
+            // fill in transposed system
+            ValueType *transposed_system = new ValueType[size * size];
```

Comment:
Miss the deconstructor. Can it use the gko::Array?
gko::matrix can not be used here due to circular dependence. Does it also apply to gko::Array?

Reply:
Good catch! No, you can use gko::Array, since it is a header-only type. I preallocated temporary storage for the triangular solve, you should be able to use the same approach here.